src/types/integer.rs

```diff
@@ -181,6 +181,40 @@ impl Neg for Integer {
     }
 }
 
+impl Add<Self> for Integer {
+    type Output = Integer;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        // The alias 'Big' differentiates the enum variant from the wrapped 'BigInt' type
+        use Integer::{BigInt as Big, I64};
+        match (self, rhs) {
+            (I64(this), I64(that)) => {
+                // Try to add the i64s together; if they overflow, upconvert to BigInts
+                match this.checked_add(that) {
+                    Some(result) => I64(result),
+                    None => Big(BigInt::from(this).add(BigInt::from(that))),
+                }
+            }
+            (I64(this), Big(that)) => Big(BigInt::from(this).add(that)),
+            (Big(this), I64(that)) => Big(this.add(&BigInt::from(that))),
+            (Big(this), Big(that)) => Big(this.add(&that)),
+        }
+    }
+}
+
+impl Zero for Integer {
```

Comment:
🗺️ There are a couple of places where we read an `Integer` and then want to find out if it's zero. Having an implementation of the `Zero` trait means that we don't have to do this check manually everywhere. The `Add` trait, implemented above, is a prerequisite trait for implementing `Zero`.

Longer-term, I think it would make sense for us to provide implementations of all the common `num` traits for `Integer` so users don't have to manually destructure each value and then do their own math operations.